entitlement/src/main/java/org/killbill/billing/entitlement/engine/core/DefaultEventsStream.java

```diff
@@ -353,73 +336,70 @@ private Collection<BlockingState> computeAddonsBlockingStatesForNextSubscription
 
     private Collection<BlockingState> computeAddonsBlockingStatesForSubscriptionBaseEvent(@Nullable final Product baseTransitionTriggerNextProduct,
                                                                                           final DateTime blockingStateEffectiveDate) {
-        if (baseSubscription == null || baseSubscription.getLastActivePlan() == null || !ProductCategory.BASE.equals(baseSubscription.getLastActivePlan().getProduct().getCategory())) {
-            return ImmutableList.<BlockingState>of();
+        if (baseSubscription == null ||
+            baseSubscription.getLastActivePlan() == null ||
+            !ProductCategory.BASE.equals(baseSubscription.getLastActivePlan().getProduct().getCategory())) {
+            return Collections.emptySet();
         }
 
+        // Retrieve all add-ons to block for that base subscription
+        final Collection<SubscriptionBase> futureBlockedAddons = allSubscriptionsForBundle
+                .stream()
+                .filter(subscription -> isAddOnsNeedToBeBlocked(subscription, baseTransitionTriggerNextProduct))
+                .collect(Collectors.toUnmodifiableList());
+
+        // Create the blocking states
+        return futureBlockedAddons.stream()
+                .map(input -> new DefaultBlockingState(input.getId(),
+                                                       BlockingStateType.SUBSCRIPTION,
+                                                       DefaultEntitlementApi.ENT_STATE_CANCELLED,
+                                                       KILLBILL_SERVICES.ENTITLEMENT_SERVICE.getServiceName(),
+                                                       true,
+                                                       true,
+                                                       false,
+                                                       blockingStateEffectiveDate))
+                .collect(Collectors.toUnmodifiableList());
+    }
+
+    /**
+     * FIXME-1615: Complex logic. Need to test properly.
```

Comment:
Following-up on the Slack discussion, I think there is room for additional unit tests in a lot of our modules.

We have lots of integration tests, either in Kill Bill (`beatrix`) or outside (https://github.com/killbill/killbill-integration-tests), which verify various features, but we could enhance our coverage on a per method / class basis. This is sometimes difficult though as the setup required to write such tests is complex.

Let's continue to think about it, maybe something we can slowly improve over time, especially as we near the 0.24 release towards the end of the year.

On a related note, addressing Spotbugs warnings and errors would be nice too (/cc @vnandwana)...

Reply:
Ok then.. Btw, lets keep that FIXME and I will back to this once `killbill-util` `collect` package refactored.